File: clients/android/services/SyncFlowService.kt
/**
 * SyncFlow Service - Android Application Integration
 *
 * High-level service that wraps SyncFlowApiClient with:
 * - Token persistence (EncryptedSharedPreferences)
 * - Auto-reconnection
 * - Background sync with WorkManager
 * - LiveData/Flow state management
 */

package com.syncflow.services

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.syncflow.api.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

// ==================== State ====================

data class SyncFlowState(
    val isAuthenticated: Boolean = false,
    val isConnected: Boolean = false,
    val isPaired: Boolean = false,
    val userId: String? = null,
    val deviceId: String? = null,
    val messages: List<Message> = emptyList(),
    val contacts: List<Contact> = emptyList(),
    val calls: List<CallHistoryEntry> = emptyList(),
    val devices: List<Device> = emptyList()
)

// ==================== Service ====================

class SyncFlowService private constructor(
    private val context: Context,
    apiUrl: String = "http://5.78.188.206"
) {
    private val client = SyncFlowApiClient(apiUrl)
    private val prefs: SharedPreferences
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val _state = MutableStateFlow(SyncFlowState())
    val state: StateFlow<SyncFlowState> = _state.asStateFlow()

    // Event flows
    private val _messageEvents = MutableSharedFlow<Message>()
    val messageEvents: SharedFlow<Message> = _messageEvents.asSharedFlow()

    private val _connectionEvents = MutableSharedFlow<Boolean>()
    val connectionEvents: SharedFlow<Boolean> = _connectionEvents.asSharedFlow()

    init {
        // Initialize encrypted preferences
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        prefs = EncryptedSharedPreferences.create(
            context,
            "syncflow_auth",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    // ==================== Initialization ====================

    suspend fun init(): Boolean {
        val accessToken = prefs.getString("accessToken", null)
        val refreshToken = prefs.getString("refreshToken", null)
        val userId = prefs.getString("userId", null)
        val deviceId = prefs.getString("deviceId", null)

        if (accessToken != null && refreshToken != null) {
            client.setTokens(accessToken, refreshToken)

            return try {
                client.getCurrentUser()
                updateState {
                    copy(
                        isAuthenticated = true,
                        isPaired = true,
                        userId = userId,
                        deviceId = deviceId
                    )
                }
                connectWebSocket()
                true
            } catch (e: Exception) {
                clearSession()
                false
            }
        }

        return false
    }

    // ==================== Authentication ====================

    suspend fun authenticateAnonymous(): User {
        val user = client.authenticateAnonymous()
        saveSession(user)
        updateState { copy(isAuthenticated = true, userId = user.userId, deviceId = user.deviceId) }
        return user
    }

    suspend fun initiatePairing(deviceName: String): String {
        val result = client.initiatePairing(deviceName, "android")
        updateState { copy(deviceId = result.deviceId) }
        return result.pairingToken
    }

    suspend fun completePairing(token: String) {
        client.completePairing(token)
    }

    suspend fun checkPairingStatus(token: String): Boolean {
        val status = client.checkPairingStatus(token)
        return status.approved
    }

    suspend fun redeemPairing(token: String): User {
        val user = client.redeemPairing(token, null, null)
        saveSession(user)
        updateState {
            copy(
                isAuthenticated = true,
                isPaired = true,
                userId = user.userId,
                deviceId = user.deviceId
            )
        }
        connectWebSocket()
        return user
    }

    fun logout() {
        client.disconnectWebSocket()
        clearSession()
        updateState { SyncFlowState() }
    }

    private fun saveSession(user: User) {
        val tokens = client.getTokens() ?: return
        prefs.edit().apply {
            putString("accessToken", tokens.accessToken)
            putString("refreshToken", tokens.refreshToken)
            putString("userId", user.userId)
            putString("deviceId", user.deviceId)
            apply()
        }
    }

    private fun clearSession() {
        prefs.edit().clear().apply()
    }

    // ==================== Messages ====================

    suspend fun loadMessages(limit: Int = 100, before: Long? = null): List<Message> {
        val response = client.getMessages(limit, before)
        val existing = _state.value.messages.associateBy { it.id }
        val merged = (response.messages + existing.values)
            .distinctBy { it.id }
            .sortedByDescending { it.date }

        updateState { copy(messages = merged) }
        return response.messages
    }

    suspend fun syncMessages(messages: List<Message>): SyncResponse {
        return client.syncMessages(messages)
    }

    suspend fun sendMessage(address: String, body: String, simSubscriptionId: Int? = null) {
        client.sendMessage(address, body, simSubscriptionId)
    }

    suspend fun getOutgoingMessages(): List<OutgoingMessage> {
        return client.getOutgoingMessages()
    }

    suspend fun updateOutgoingStatus(id: String, status: String, error: String? = null) {
        client.updateOutgoingStatus(id, status, error)
    }

    suspend fun markAsRead(messageId: String) {
        client.markMessageRead(messageId)
        updateState {
            copy(messages = messages.map {
                if (it.id == messageId) it.copy(read = true) else it
            })
        }
    }

    // ==================== Contacts ====================

    suspend fun loadContacts(): List<Contact> {
        val response = client.getContacts()
        updateState { copy(contacts = response.contacts) }
        return response.contacts
    }

    suspend fun syncContacts(contacts: List<Contact>): SyncResponse {
        return client.syncContacts(contacts)
    }

    suspend fun searchContacts(query: String): List<Contact> {
        return client.getContacts(query).contacts
    }

    // ==================== Call History ====================

    suspend fun loadCallHistory(limit: Int = 100, before: Long? = null): List<CallHistoryEntry> {
        val response = client.getCallHistory(limit, before)
        val existing = _state.value.calls.associateBy { it.id }
        val merged = (response.calls + existing.values)
            .distinctBy { it.id }
            .sortedByDescending { it.callDate }

        updateState { copy(calls = merged) }
        return response.calls
    }

    suspend fun syncCallHistory(calls: List<CallHistoryEntry>): SyncResponse {
        return client.syncCallHistory(calls)
    }

    suspend fun getCallRequests(): List<CallRequest> {
        return client.getCallRequests()
    }

    suspend fun updateCallRequestStatus(id: String, status: String) {
        client.updateCallRequestStatus(id, status)
    }

    // ==================== Devices ====================

    suspend fun loadDevices(): List<Device> {
        val response = client.getDevices()
        updateState { copy(devices = response.devices) }
        return response.devices
    }

    suspend fun updateDevice(id: String, name: String? = null, fcmToken: String? = null) {
        client.updateDevice(id, name, fcmToken)
    }

    suspend fun removeDevice(id: String) {
        client.removeDevice(id)
        updateState { copy(devices = devices.filter { it.id != id }) }
    }

    // ==================== WebSocket ====================

    private fun connectWebSocket() {
        client.connectWebSocket(object : SyncFlowWebSocketListener {
            override fun onConnected() {
                scope.launch {
                    updateState { copy(isConnected = true) }
                    _connectionEvents.emit(true)
                }
                client.subscribe("messages")
                client.subscribe("contacts")
                client.subscribe("calls")
                client.subscribe("devices")
                client.subscribe("outgoing")
                client.subscribe("call_requests")
            }

            override fun onDisconnected() {
                scope.launch {
                    updateState { copy(isConnected = false) }
                    _connectionEvents.emit(false)
                }
                // Auto-reconnect after delay
                scope.launch {
                    delay(3000)
                    if (_state.value.isAuthenticated) {
                        connectWebSocket()
                    }
                }
            }

            override fun onError(error: String) {
                // Log error
            }

            override fun onMessageAdded(message: Message) {
                scope.launch {
                    val current = _state.value.messages
                    if (current.none { it.id == message.id }) {
                        updateState {
                            copy(messages = listOf(message) + messages)
                        }
                        _messageEvents.emit(message)
                    }
                }
            }

            override fun onMessageUpdated(message: Message) {
                scope.launch {
                    updateState {
                        copy(messages = messages.map {
                            if (it.id == message.id) message else it
                        })
                    }
                }
            }

            override fun onMessageDeleted(messageId: String) {
                scope.launch {
                    updateState {
                        copy(messages = messages.filter { it.id != messageId })
                    }
                }
            }

            override fun onContactAdded(contact: Contact) {
                scope.launch {
                    val current = _state.value.contacts
                    if (current.none { it.id == contact.id }) {
                        updateState { copy(contacts = contacts + contact) }
                    }
                }
            }

            override fun onContactUpdated(contact: Contact) {
                scope.launch {
                    updateState {
                        copy(contacts = contacts.map {
                            if (it.id == contact.id) contact else it
                        })
                    }
                }
            }

            override fun onContactDeleted(contactId: String) {
                scope.launch {
                    updateState {
                        copy(contacts = contacts.filter { it.id != contactId })
                    }
                }
            }

            override fun onCallAdded(call: CallHistoryEntry) {
                scope.launch {
                    val current = _state.value.calls
                    if (current.none { it.id == call.id }) {
                        updateState { copy(calls = listOf(call) + calls) }
                    }
                }
            }

            override fun onOutgoingMessage(message: OutgoingMessage) {
                // Handle outgoing message - Android should process and send
            }

            override fun onCallRequest(request: CallRequest) {
                // Handle call request - Android should initiate call
            }
        })
    }

    fun disconnectWebSocket() {
        client.disconnectWebSocket()
    }

    // ==================== State Helpers ====================

    private inline fun updateState(update: SyncFlowState.() -> SyncFlowState) {
        _state.value = _state.value.update()
    }

    // ==================== Singleton ====================

    companion object {
        @Volatile
        private var instance: SyncFlowService? = null

        fun getInstance(context: Context): SyncFlowService {
            return instance ?: synchronized(this) {
                instance ?: SyncFlowService(context.applicationContext).also {
                    instance = it
                }
            }
        }
    }
}

File: clients/macos/services/SyncFlowService.swift
/**
 * SyncFlow Service - macOS Application Integration
 *
 * High-level service that wraps SyncFlowApiClient with:
 * - Token persistence (Keychain)
 * - Auto-reconnection
 * - Combine publishers for state management
 */

import Foundation
import Combine
import Security

// MARK: - Keychain Helper

private class KeychainHelper {
    static let service = "com.syncflow.vps"

    static func save(_ value: String, forKey key: String) {
        let data = value.data(using: .utf8)!

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]

        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }

    static func get(_ key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        if status == errSecSuccess, let data = result as? Data {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }

    static func delete(_ key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(query as CFDictionary)
    }

    static func clear() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service
        ]
        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - State

public struct SyncFlowState {
    public var isAuthenticated: Bool = false
    public var isConnected: Bool = false
    public var isPaired: Bool = false
    public var userId: String?
    public var deviceId: String?
    public var messages: [Message] = []
    public var contacts: [Contact] = []
    public var calls: [CallHistoryEntry] = []
    public var devices: [Device] = []
}

// MARK: - Service

public class SyncFlowService: ObservableObject, SyncFlowWebSocketDelegate {
    public static let shared = SyncFlowService()

    private let client: SyncFlowApiClient
    private var reconnectTask: Task<Void, Never>?

    @Published public private(set) var state = SyncFlowState()

    // Event publishers
    public let messageReceived = PassthroughSubject<Message, Never>()
    public let connectionChanged = PassthroughSubject<Bool, Never>()

    private init(apiUrl: String = "http://5.78.188.206") {
        self.client = SyncFlowApiClient(baseUrl: apiUrl)
        self.client.webSocketDelegate = self
    }

    // MARK: - Initialization

    @MainActor
    public func initialize() async -> Bool {
        guard let accessToken = KeychainHelper.get("accessToken"),
              let refreshToken = KeychainHelper.get("refreshToken") else {
            return false
        }

        client.setTokens(accessToken: accessToken, refreshToken: refreshToken)
        state.userId = KeychainHelper.get("userId")
        state.deviceId = KeychainHelper.get("deviceId")

        do {
            _ = try await client.getCurrentUser()
            state.isAuthenticated = true
            state.isPaired = true
            connectWebSocket()
            return true
        } catch {
            clearSession()
            return false
        }
    }

    // MARK: - Authentication

    @MainActor
    public func authenticateAnonymous() async throws -> User {
        let user = try await client.authenticateAnonymous()
        saveSession(user: user)
        state.isAuthenticated = true
        state.userId = user.userId
        state.deviceId = user.deviceId
        return user
    }

    @MainActor
    public func initiatePairing(deviceName: String) async throws -> String {
        let result = try await client.initiatePairing(deviceName: deviceName, deviceType: "macos")
        state.deviceId = result.deviceId
        return result.pairingToken
    }

    public func checkPairingStatus(token: String) async throws -> Bool {
        let status = try await client.checkPairingStatus(token: token)
        return status.approved
    }

    @MainActor
    public func redeemPairing(token: String) async throws -> User {
        let user = try await client.redeemPairing(token: token, deviceName: nil, deviceType: "macos")
        saveSession(user: user)
        state.isAuthenticated = true
        state.isPaired = true
        state.userId = user.userId
        state.deviceId = user.deviceId
        connectWebSocket()
        return user
    }

    @MainActor
    public func logout() {
        client.disconnectWebSocket()
        clearSession()
        state = SyncFlowState()
    }

    private func saveSession(user: User) {
        guard let tokens = client.getTokens() else { return }
        KeychainHelper.save(tokens.accessToken, forKey: "accessToken")
        KeychainHelper.save(tokens.refreshToken, forKey: "refreshToken")
        KeychainHelper.save(user.userId, forKey: "userId")
        KeychainHelper.save(user.deviceId, forKey: "deviceId")
    }

    private func clearSession() {
        KeychainHelper.clear()
    }

    // MARK: - Messages

    @MainActor
    public func loadMessages(limit: Int = 100, before: Int64? = nil) async throws -> [Message] {
        let response = try await client.getMessages(limit: limit, before: before)
        let existingIds = Set(state.messages.map { $0.id })
        let newMessages = response.messages.filter { !existingIds.contains($0.id) }
        state.messages = (state.messages + newMessages).sorted { $0.date > $1.date }
        return response.messages
    }

    public func syncMessages(_ messages: [[String: Any]]) async throws -> SyncResponse {
        return try await client.syncMessages(messages)
    }

    public func sendMessage(address: String, body: String, simSubscriptionId: Int? = nil) async throws {
        _ = try await client.sendMessage(address: address, body: body, simSubscriptionId: simSubscriptionId)
    }

    @MainActor
    public func markAsRead(messageId: String) async throws {
        try await client.markMessageRead(id: messageId)
        if let index = state.messages.firstIndex(where: { $0.id == messageId }) {
            // Create updated message with read = true
            let oldMsg = state.messages[index]
            let updatedMsg = Message(
                id: oldMsg.id,
                threadId: oldMsg.threadId,
                address: oldMsg.address,
                contactName: oldMsg.contactName,
                body: oldMsg.body,
                date: oldMsg.date,
                type: oldMsg.type,
                read: true,
                isMms: oldMsg.isMms,
                encrypted: oldMsg.encrypted
            )
            state.messages[index] = updatedMsg
        }
    }

    // MARK: - Contacts

    @MainActor
    public func loadContacts() async throws -> [Contact] {
        let response = try await client.getContacts()
        state.contacts = response.contacts
        return response.contacts
    }

    public func syncContacts(_ contacts: [[String: Any]]) async throws -> SyncResponse {
        return try await client.syncContacts(contacts)
    }

    public func searchContacts(query: String) async throws -> [Contact] {
        let response = try await client.getContacts(search: query)
        return response.contacts
    }

    // MARK: - Call History

    @MainActor
    public func loadCallHistory(limit: Int = 100, before: Int64? = nil) async throws -> [CallHistoryEntry] {
        let response = try await client.getCallHistory(limit: limit, before: before)
        let existingIds = Set(state.calls.map { $0.id })
        let newCalls = response.calls.filter { !existingIds.contains($0.id) }
        state.calls = (state.calls + newCalls).sorted { $0.callDate > $1.callDate }
        return response.calls
    }

    public func syncCallHistory(_ calls: [[String: Any]]) async throws -> SyncResponse {
        return try await client.syncCallHistory(calls)
    }

    public func makeCall(phoneNumber: String, simSubscriptionId: Int? = nil) async throws {
        _ = try await client.requestCall(phoneNumber: phoneNumber, simSubscriptionId: simSubscriptionId)
    }

    // MARK: - Devices

    @MainActor
    public func loadDevices() async throws -> [Device] {
        let response = try await client.getDevices()
        state.devices = response.devices
        return response.devices
    }

    @MainActor
    public func removeDevice(id: String) async throws {
        try await client.removeDevice(id: id)
        state.devices = state.devices.filter { $0.id != id }
    }

    // MARK: - WebSocket

    private func connectWebSocket() {
        client.connectWebSocket()
        client.subscribe("messages")
        client.subscribe("contacts")
        client.subscribe("calls")
        client.subscribe("devices")
    }

    private func scheduleReconnect() {
        reconnectTask?.cancel()
        reconnectTask = Task {
            try? await Task.sleep(nanoseconds: 3_000_000_000) // 3 seconds
            if state.isAuthenticated && !Task.isCancelled {
                connectWebSocket()
            }
        }
    }

    // MARK: - WebSocket Delegate

    public func onConnected() {
        DispatchQueue.main.async {
            self.state.isConnected = true
            self.connectionChanged.send(true)
        }
    }

    public func onDisconnected() {
        DispatchQueue.main.async {
            self.state.isConnected = false
            self.connectionChanged.send(false)
        }
        scheduleReconnect()
    }

    public func onError(_ error: Error) {
        print("WebSocket error: \(error)")
    }

    public func onMessageAdded(_ message: Message) {
        DispatchQueue.main.async {
            if !self.state.messages.contains(where: { $0.id == message.id }) {
                self.state.messages.insert(message, at: 0)
                self.messageReceived.send(message)
            }
        }
    }

    public func onMessageUpdated(_ message: Message) {
        DispatchQueue.main.async {
            if let index = self.state.messages.firstIndex(where: { $0.id == message.id }) {
                self.state.messages[index] = message
            }
        }
    }

    public func onMessageDeleted(_ messageId: String) {
        DispatchQueue.main.async {
            self.state.messages.removeAll { $0.id == messageId }
        }
    }

    public func onContactAdded(_ contact: Contact) {
        DispatchQueue.main.async {
            if !self.state.contacts.contains(where: { $0.id == contact.id }) {
                self.state.contacts.append(contact)
            }
        }
    }

    public func onContactUpdated(_ contact: Contact) {
        DispatchQueue.main.async {
            if let index = self.state.contacts.firstIndex(where: { $0.id == contact.id }) {
                self.state.contacts[index] = contact
            }
        }
    }

    public func onContactDeleted(_ contactId: String) {
        DispatchQueue.main.async {
            self.state.contacts.removeAll { $0.id == contactId }
        }
    }

    public func onCallAdded(_ call: CallHistoryEntry) {
        DispatchQueue.main.async {
            if !self.state.calls.contains(where: { $0.id == call.id }) {
                self.state.calls.insert(call, at: 0)
            }
        }
    }

    public func onOutgoingMessage(_ message: OutgoingMessage) {
        // Handle on Android side
    }

    public func onCallRequest(_ request: CallRequest) {
        // Handle on Android side
    }
}
